File: src/problemsolvingpatterns/multipepointers/sumzero/SumZero.js
/*
    Write a function called sumZero which accepts a sorted array of integers.
    The function should find the first pair where the sum is 0.
    Return an array that includes both values that sum to zero or undefined
    if a pair does not exist

sumZero([-3,-2,-1,0,1,2,3]) // [-3,3] 
sumZero([-2,0,1,3]) // undefined
sumZero([1,2,3]) // undefined

1-Solution O(n**2)

// loop over the array
    // define placeholdr variable to hold the value of the current index
    // loop over the array from the current index
    // sum the each value from the inner loop
    // is the sum is zero, return the pair as an array
// return undefined

2-Solution O(n)

// define the two pointer
// while substruction of two pointers are greater than zero
    // sum the value of the pointers
    // if the sum is less than zero, move the left pointer to left
    // if the sum is greater than zero, move the right pointer to the right
    else return the pair as an array
// return undefined
*/

// 1-Solution O(n**2)
function sumZero(arr) {
  for (let i = 0; i < arr.length; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      let sum = arr[i] + arr[j];
      if (sum === 0) {
        return [arr[i], arr[j]]
      }
    }
  }
  return undefined;
}

// 2-Solution O(n)
function sumZero(arr) {
  if (arr.length <= 1) {
    return undefined;
  }
  let left = 0;
  let right = arr.length - 1;
  while ((left - right) !== 0) {
    let sum = arr[left] + arr[right];
    if (sum > 0) {
      right--;
    } else if (sum < 0) {
      left++;
    } else {
      return [arr[left], arr[right]];
    }
  }
  return undefined;
}

sumZero([-3, -2, -1, 0, 1, 2, 3])

File: src/problemsolvingpatterns/multipepointers/sumzero/SumZero.java
package problemsolvingpatterns.multipepointers.sumzero;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Write a function called sumZero which accepts a sorted array of integers.
 * The function should find the first pair where the sum is 0.
 * Return an array that includes both values that sum to zero or undefined
 * if a pair does not exist
 * 
 * Examples: 
 * 
 * sumZero([-3,-2,-1,0,1,2,3]) // [-3,3] 
 * sumZero([-2,0,1,3]) // undefined
 * sumZero([1,2,3]) // undefined
 * 
 */
public class SumZero {
  /**
   * Approach #1: O(n^2)
   * 
   * loop over the each element and compare each element with other element in the nested loop
   * 
   * @param numbers
   * @return
   */
  public static int[] sumZero1(int[] numbers) {
    for (int i = 0; i < numbers.length; i++) {
      for (int j = i + 1; j < numbers.length; j++) {
        int sum = numbers[i] + numbers[j];
        if (sum == 0) {
          return new int[]{numbers[i], numbers[i]};
        }
      }
    }
    return new int[]{};
  }

  /**
   * Approach #2: O(n)
   * 
   * define a two pointers left and right. Assign the first index to the left and the last to the right pointers
   * loop over the array
   *  sum the elements at left and right
   *  if the sum is zero return new array containing the element at the left and right pointers
   *  if the sum is less than zero, move the left pointer to the right by one
   *  if the sum is greater than zero, move the right pointer to the left by one
   * end the loop
   * return empty array
   * 
   * @return
   */
  public static int[] sumZero2(int[] numbers) {
    int left = 0; 
    int right = numbers.length - 1;

    while (left != right) {
      int sum = numbers[left] + numbers[right];
      if (sum == 0) {
        return new int[]{numbers[left], numbers[right]};
      } else if (sum < 0) {
        left++;
      } else {
        right--;
      }
    }
    return new int[]{};
  }

  public static void main(String[] args) {
    test(new int[]{-3,-2,-1,0,1,2,3}, 1);
    test(new int[]{-2,0,1,3}, 2);
    test(new int[]{1,2,3}, 3);
    test(new int[]{-4,-2,-1,0,1,2,3}, 4);
  }

  public static void test(int[] nums, int caseNumber) {
    System.out.println("Case #" + caseNumber + ": " + Arrays.stream(sumZero2(nums)).boxed().toList());
  }
}
